In [ ]:
import pandas as pd

# Companies House Company Data Exploration

In [16]:
df = pd.read_csv(
    "data/companies/BasicCompanyData-2026-08-01-part1_7.csv",
    low_memory=False,
)

## Drop columns and basic cleaning

In [17]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 849999 entries, 0 to 849998
Data columns (total 55 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   CompanyName                         849999 non-null  str    
 1    CompanyNumber                      849999 non-null  str    
 2   RegAddress.CareOf                   4228 non-null    str    
 3   RegAddress.POBox                    4673 non-null    str    
 4   RegAddress.AddressLine1             842520 non-null  str    
 5    RegAddress.AddressLine2            500803 non-null  str    
 6   RegAddress.PostTown                 836228 non-null  str    
 7   RegAddress.County                   240693 non-null  str    
 8   RegAddress.Country                  710148 non-null  str    
 9   RegAddress.PostCode                 838187 non-null  str    
 10  CompanyCategory                     849999 non-null  str    
 11  CompanyStatus                       8

In [18]:
df.columns

Index(['CompanyName', ' CompanyNumber', 'RegAddress.CareOf',
       'RegAddress.POBox', 'RegAddress.AddressLine1',
       ' RegAddress.AddressLine2', 'RegAddress.PostTown', 'RegAddress.County',
       'RegAddress.Country', 'RegAddress.PostCode', 'CompanyCategory',
       'CompanyStatus', 'CountryOfOrigin', 'DissolutionDate',
       'IncorporationDate', 'Accounts.AccountRefDay',
       'Accounts.AccountRefMonth', 'Accounts.NextDueDate',
       'Accounts.LastMadeUpDate', 'Accounts.AccountCategory',
       'Returns.NextDueDate', 'Returns.LastMadeUpDate',
       'Mortgages.NumMortCharges', 'Mortgages.NumMortOutstanding',
       'Mortgages.NumMortPartSatisfied', 'Mortgages.NumMortSatisfied',
       'SICCode.SicText_1', 'SICCode.SicText_2', 'SICCode.SicText_3',
       'SICCode.SicText_4', 'LimitedPartnerships.NumGenPartners',
       'LimitedPartnerships.NumLimPartners', 'URI', 'PreviousName_1.CONDATE',
       ' PreviousName_1.CompanyName', ' PreviousName_2.CONDATE',
       ' PreviousName_2.C

### Drop columns

In [19]:
previous_name_cols = [
    col for col in df.columns
    if "PreviousName_" in col
]
previous_name_cols

['PreviousName_1.CONDATE',
 ' PreviousName_1.CompanyName',
 ' PreviousName_2.CONDATE',
 ' PreviousName_2.CompanyName',
 'PreviousName_3.CONDATE',
 ' PreviousName_3.CompanyName',
 'PreviousName_4.CONDATE',
 ' PreviousName_4.CompanyName',
 'PreviousName_5.CONDATE',
 ' PreviousName_5.CompanyName',
 'PreviousName_6.CONDATE',
 ' PreviousName_6.CompanyName',
 'PreviousName_7.CONDATE',
 ' PreviousName_7.CompanyName',
 'PreviousName_8.CONDATE',
 ' PreviousName_8.CompanyName',
 'PreviousName_9.CONDATE',
 ' PreviousName_9.CompanyName',
 'PreviousName_10.CONDATE',
 ' PreviousName_10.CompanyName']

In [20]:
# Previous names are not useful for our purposes

df = df.drop(columns=previous_name_cols)

In [21]:
df.columns

Index(['CompanyName', ' CompanyNumber', 'RegAddress.CareOf',
       'RegAddress.POBox', 'RegAddress.AddressLine1',
       ' RegAddress.AddressLine2', 'RegAddress.PostTown', 'RegAddress.County',
       'RegAddress.Country', 'RegAddress.PostCode', 'CompanyCategory',
       'CompanyStatus', 'CountryOfOrigin', 'DissolutionDate',
       'IncorporationDate', 'Accounts.AccountRefDay',
       'Accounts.AccountRefMonth', 'Accounts.NextDueDate',
       'Accounts.LastMadeUpDate', 'Accounts.AccountCategory',
       'Returns.NextDueDate', 'Returns.LastMadeUpDate',
       'Mortgages.NumMortCharges', 'Mortgages.NumMortOutstanding',
       'Mortgages.NumMortPartSatisfied', 'Mortgages.NumMortSatisfied',
       'SICCode.SicText_1', 'SICCode.SicText_2', 'SICCode.SicText_3',
       'SICCode.SicText_4', 'LimitedPartnerships.NumGenPartners',
       'LimitedPartnerships.NumLimPartners', 'URI', 'ConfStmtNextDueDate',
       ' ConfStmtLastMadeUpDate'],
      dtype='str')

In [22]:
df["URI"].head()

0    http://business.data.gov.uk/id/company/08209948
1    http://business.data.gov.uk/id/company/16092999
2    http://business.data.gov.uk/id/company/11743365
3    http://business.data.gov.uk/id/company/16873705
4    http://business.data.gov.uk/id/company/15073164
Name: URI, dtype: str

In [23]:
# URI is derivable from company number -> drop

df = df.drop(columns=["URI"])

In [25]:
df.columns

Index(['CompanyName', ' CompanyNumber', 'RegAddress.CareOf',
       'RegAddress.POBox', 'RegAddress.AddressLine1',
       ' RegAddress.AddressLine2', 'RegAddress.PostTown', 'RegAddress.County',
       'RegAddress.Country', 'RegAddress.PostCode', 'CompanyCategory',
       'CompanyStatus', 'CountryOfOrigin', 'DissolutionDate',
       'IncorporationDate', 'Accounts.AccountRefDay',
       'Accounts.AccountRefMonth', 'Accounts.NextDueDate',
       'Accounts.LastMadeUpDate', 'Accounts.AccountCategory',
       'Returns.NextDueDate', 'Returns.LastMadeUpDate',
       'Mortgages.NumMortCharges', 'Mortgages.NumMortOutstanding',
       'Mortgages.NumMortPartSatisfied', 'Mortgages.NumMortSatisfied',
       'SICCode.SicText_1', 'SICCode.SicText_2', 'SICCode.SicText_3',
       'SICCode.SicText_4', 'LimitedPartnerships.NumGenPartners',
       'LimitedPartnerships.NumLimPartners', 'ConfStmtNextDueDate',
       ' ConfStmtLastMadeUpDate'],
      dtype='str')

### Clean column names

In [26]:
df.columns = df.columns.str.strip()

In [27]:
df.columns

Index(['CompanyName', 'CompanyNumber', 'RegAddress.CareOf', 'RegAddress.POBox',
       'RegAddress.AddressLine1', 'RegAddress.AddressLine2',
       'RegAddress.PostTown', 'RegAddress.County', 'RegAddress.Country',
       'RegAddress.PostCode', 'CompanyCategory', 'CompanyStatus',
       'CountryOfOrigin', 'DissolutionDate', 'IncorporationDate',
       'Accounts.AccountRefDay', 'Accounts.AccountRefMonth',
       'Accounts.NextDueDate', 'Accounts.LastMadeUpDate',
       'Accounts.AccountCategory', 'Returns.NextDueDate',
       'Returns.LastMadeUpDate', 'Mortgages.NumMortCharges',
       'Mortgages.NumMortOutstanding', 'Mortgages.NumMortPartSatisfied',
       'Mortgages.NumMortSatisfied', 'SICCode.SicText_1', 'SICCode.SicText_2',
       'SICCode.SicText_3', 'SICCode.SicText_4',
       'LimitedPartnerships.NumGenPartners',
       'LimitedPartnerships.NumLimPartners', 'ConfStmtNextDueDate',
       'ConfStmtLastMadeUpDate'],
      dtype='str')

## Load SIC Codes

In [28]:
sic_df = pd.read_csv("data/sic-codes/SIC07_CH_condensed_list_en.csv")

sic_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   SIC Code     731 non-null    int64
 1   Description  731 non-null    str  
dtypes: int64(1), str(1)
memory usage: 11.6 KB


In [29]:
sic_df.head()

,SIC Code,Description
0,1110,"Growing of cereals (except rice), leguminous c..."
1,1120,Growing of rice
2,1130,"Growing of vegetables and melons, roots and tu..."
3,1140,Growing of sugar cane
4,1150,Growing of tobacco


In [30]:
# SIC code should be string not int

sic_df = pd.read_csv(
    "data/sic-codes/SIC07_CH_condensed_list_en.csv",
    dtype={"SIC Code": "string"},
)


In [31]:
sic_df.head()

,SIC Code,Description
0,01110,"Growing of cereals (except rice), leguminous c..."
1,01120,Growing of rice
2,01130,"Growing of vegetables and melons, roots and tu..."
3,01140,Growing of sugar cane
4,01150,Growing of tobacco


## Inspect Company Data SIC Code Frequency

In [34]:
df[
    [
        "SICCode.SicText_1",
        "SICCode.SicText_2",
        "SICCode.SicText_3",
        "SICCode.SicText_4", 
    ]
].head(10)

,SICCode.SicText_1,SICCode.SicText_2,SICCode.SicText_3,SICCode.SicText_4
0,99999 - Dormant Company,NaN,NaN,NaN
1,62020 - Information technology consultancy act...,NaN,NaN,NaN
2,59112 - Video production activities,63120 - Web portals,74100 - specialised design activities,74201 - Portrait photographic activities
3,86210 - General medical practice activities,NaN,NaN,NaN
4,70229 - Management consultancy activities othe...,NaN,NaN,NaN
5,58290 - Other software publishing,NaN,NaN,NaN
6,62090 - Other information technology service a...,70229 - Management consultancy activities othe...,NaN,NaN
7,68209 - Other letting and operating of own or ...,NaN,NaN,NaN
8,70229 - Management consultancy activities othe...,NaN,NaN,NaN
9,None Supplied,NaN,NaN,NaN


### Value Counts

In [35]:
df["SICCode.SicText_1"].value_counts().head(20)

SICCode.SicText_1
68209 - Other letting and operating of own or leased real estate             44484
68100 - Buying and selling of own real estate                                41739
98000 - Residents property management                                        38717
None Supplied                                                                34391
70229 - Management consultancy activities other than financial management    31874
82990 - Other business support service activities n.e.c.                     26078
47910 - Retail sale via mail order houses or via Internet                    23167
62020 - Information technology consultancy activities                        18756
41100 - Development of building projects                                     17833
64209 - Activities of other holding companies n.e.c.                         17557
96090 - Other service activities n.e.c.                                      17442
99999 - Dormant Company                                              

In [36]:
df["SICCode.SicText_2"].value_counts().head(20)

SICCode.SicText_2
68209 - Other letting and operating of own or leased real estate             23924
70229 - Management consultancy activities other than financial management     6321
62020 - Information technology consultancy activities                         5112
41202 - Construction of domestic buildings                                    4499
47910 - Retail sale via mail order houses or via Internet                     4486
82990 - Other business support service activities n.e.c.                      3618
68100 - Buying and selling of own real estate                                 3539
62090 - Other information technology service activities                       3484
68320 - Management of real estate on a fee or contract basis                  2960
96090 - Other service activities n.e.c.                                       2737
68201 - Renting and operating of Housing Association real estate              2598
69202 - Bookkeeping activities                                       

In [37]:
df["SICCode.SicText_3"].value_counts().head(20)

SICCode.SicText_3
68320 - Management of real estate on a fee or contract basis                 6327
68209 - Other letting and operating of own or leased real estate             4563
70229 - Management consultancy activities other than financial management    2944
62090 - Other information technology service activities                      2228
82990 - Other business support service activities n.e.c.                     2144
47910 - Retail sale via mail order houses or via Internet                    2030
96090 - Other service activities n.e.c.                                      1542
62020 - Information technology consultancy activities                        1264
85590 - Other education n.e.c.                                               1204
73110 - Advertising agencies                                                 1150
69203 - Tax consultancy                                                      1088
90030 - Artistic creation                                                    103

In [38]:
df["SICCode.SicText_4"].value_counts().head(20)

SICCode.SicText_4
68320 - Management of real estate on a fee or contract basis                 2989
82990 - Other business support service activities n.e.c.                     1583
96090 - Other service activities n.e.c.                                      1358
70229 - Management consultancy activities other than financial management    1337
47910 - Retail sale via mail order houses or via Internet                    1092
90030 - Artistic creation                                                     909
85600 - Educational support services                                          776
85590 - Other education n.e.c.                                                764
74100 - specialised design activities                                         659
74909 - Other professional, scientific and technical activities n.e.c.        638
68209 - Other letting and operating of own or leased real estate              628
73110 - Advertising agencies                                                  61

### Question 1:

**Across all SIC classifications supplied by companies, which activities occur most frequently?**

In [39]:
sic_cols = [
    "SICCode.SicText_1",
    "SICCode.SicText_2",
    "SICCode.SicText_3",
    "SICCode.SicText_4",
]

all_sics = df[sic_cols].stack()

all_sics.value_counts().head(20)

68209 - Other letting and operating of own or leased real estate             73599
68100 - Buying and selling of own real estate                                46411
70229 - Management consultancy activities other than financial management    42476
98000 - Residents property management                                        41112
None Supplied                                                                34391
82990 - Other business support service activities n.e.c.                     33423
47910 - Retail sale via mail order houses or via Internet                    30775
62020 - Information technology consultancy activities                        25400
96090 - Other service activities n.e.c.                                      23079
68320 - Management of real estate on a fee or contract basis                 22384
64209 - Activities of other holding companies n.e.c.                         18204
41100 - Development of building projects                                     18036
9999

In [ ]:
all_sics.nunique()

1042

### Check SIC code format

Expect: all 5 digits

In [41]:
sic_df["SIC Code"].str.len().value_counts()

SIC Code
5    731
Name: count, dtype: Int64

In [42]:
sic_df["SIC Code"].str.isdigit().value_counts()

SIC Code
True    731
Name: count, dtype: Int64

### Check Company Data SIC Codes are all 5 digits

In [47]:
all_sics = df[sic_cols].stack().dropna()

sic_parts = all_sics.str.split(" - ", n=1, expand=True)
sic_parts.columns = ["SICCode", "Description"]

sic_parts["SICCode"] = sic_parts["SICCode"].str.strip()
sic_parts["Description"] = sic_parts["Description"].str.strip()

sic_parts.head()

SICCode                                    Description
0 SICCode.SicText_1   99999                                Dormant Company
1 SICCode.SicText_1   62020  Information technology consultancy activities
2 SICCode.SicText_1   59112                    Video production activities
  SICCode.SicText_2   63120                                    Web portals
  SICCode.SicText_3   74100                  specialised design activities

In [48]:
sic_parts["SICCode"].str.len().value_counts()

SICCode
5     1093500
13      34391
4        2981
Name: count, dtype: int64

In [49]:
sic_parts["SICCode"].str.isdigit().value_counts()

SICCode
True     1096481
False      34391
Name: count, dtype: int64

### Question 2:

**Why do we have 2981 four-digit numeric codes**

In [50]:
sic_parts.loc[
    sic_parts["SICCode"].str.len() == 4,
    "SICCode"
].value_counts()

SICCode
7499    314
4521    283
7487    254
7011     98
4525     87
       ... 
7483      1
5523      1
2912      1
5181      1
6021      1
Name: count, Length: 312, dtype: int64

In [51]:
(
    sic_parts.loc[
        sic_parts["SICCode"].str.len() == 4,
        ["SICCode", "Description"]
    ]
    .drop_duplicates()
    .sort_values("SICCode")
)

SICCode  \
275950 SICCode.SicText_1    0111   
301781 SICCode.SicText_1    0112   
       SICCode.SicText_2    0122   
1043   SICCode.SicText_1    0130   
222528 SICCode.SicText_1    0141   
...                          ...   
997    SICCode.SicText_1    9305   
6816   SICCode.SicText_2    9500   
39018  SICCode.SicText_1    9600   
660917 SICCode.SicText_1    9800   
821    SICCode.SicText_1    9999   

                                                                Description  
275950 SICCode.SicText_1                         Grow cereals & other crops  
301781 SICCode.SicText_1                 Grow vegetables & nursery products  
       SICCode.SicText_2                    Farm sheep, goats, horses, etc.  
1043   SICCode.SicText_1           Crops combined with animals, mixed farms  
222528 SICCode.SicText_1                    Agricultural service activities  
...                                                                     ...  
997    SICCode.SicText_1                    Other service activities n.e.c.  
6816   SICCode.SicText_2                  Private households with employees  
39018  SICCode.SicText_1  Undifferentiated goods producing activities of...  
660917 SICCode.SicText_1                      Residents property management  
821    SICCode.SicText_1                                    Dormant company  

[312 rows x 2 columns]